metricbeat/helper/kubernetes/state_metricset.go

```diff
@@ -90,6 +97,17 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
 // format. It publishes the event which is then forwarded to the output. In case
 // of an error set the Error field of mb.Event or simply call report.Error().
 func (m *MetricSet) Fetch(reporter mb.ReporterV2) {
+	// The name of the metric state can be obtained by using m.BaseMetricSet.Name(). However, names that start with state_* (e.g. state_cronjob)
+	// need to have that prefix removed. So, for example, strings.ReplaceAll("state_cronjob", "state_", "") would result in just cronjob.
+	// Exception is state_namespace, as field kubernetes.namespace already exists, and we need to create a new object
+	// for the state_namespace metricset.
+	resourceName := m.BaseMetricSet.Name()
+	if resourceName != util.NamespaceResource {
+		resourceName = strings.ReplaceAll(resourceName, prefix, "")
+	} else {
+		resourceName = "namespace_state"
```

Comment:
I think the resource name should be `namespace` so as to be used in `event.dataset` properly. Potentially we could extend the metricset to also add other fields like `namespace_foo` so just using `namespace_state` for the dataset I think is not so correct.

Reply:
I am afraid that will break again because `kubernetes.state` already exists :thinking:

Reply:
I mean that the value of `event.dataset` should be set to `namespace`. I think this would work since it's the value that we set so I don't see how it would be in conflict with anything else, right?

Reply:
It would create a conflict, because that field is setting the object `kubernetes.namespace_state`. And if we change it to `kubernetes.namespace`, this is the document that is generated (I just ran this):

```json
[
    {
        "event": {
            "dataset": "kubernetes.namespace",
            "duration": 115000,
            "module": "kubernetes"
        },
        "kubernetes": {
            "namespace": {
                "created": {
                    "sec": 1691566342
                },
                "status": {
                    "active": true,
                    "terminating": false
                }
            }
        },
        "metricset": {
            "name": "state_namespace",
            "period": 10000
        },
        "service": {
            "address": "127.0.0.1:55555",
            "type": "kubernetes"
        }
    },
    {
        "event": {
            "dataset": "kubernetes.namespace",
            "duration": 115000,
            "module": "kubernetes"
        },
        "kubernetes": {
            "namespace": {
                "created": {
                    "sec": 1691566338
                },
                "status": {
                    "active": true,
                    "terminating": false
                }
            }
        },
        "metricset": {
            "name": "state_namespace",
            "period": 10000
        },
        "service": {
            "address": "127.0.0.1:55555",
            "type": "kubernetes"
        }
    },
    {
        "event": {
            "dataset": "kubernetes.namespace",
            "duration": 115000,
            "module": "kubernetes"
        },
        "kubernetes": {
            "namespace": {
                "created": {
                    "sec": 1691566337
                },
                "status": {
                    "active": true,
                    "terminating": false
                }
            }
        },
        "metricset": {
            "name": "state_namespace",
            "period": 10000
        },
        "service": {
            "address": "127.0.0.1:55555",
            "type": "kubernetes"
        }
    },
    {
        "event": {
            "dataset": "kubernetes.namespace",
            "duration": 115000,
            "module": "kubernetes"
        },
        "kubernetes": {
            "namespace": {
                "created": {
                    "sec": 1691566337
                },
                "status": {
                    "active": true,
                    "terminating": false
                }
            }
        },
        "metricset": {
            "name": "state_namespace",
            "period": 10000
        },
        "service": {
            "address": "127.0.0.1:55555",
            "type": "kubernetes"
        }
    },
    {
        "event": {
            "dataset": "kubernetes.namespace",
            "duration": 115000,
            "module": "kubernetes"
        },
        "kubernetes": {
            "namespace": {
                "created": {
                    "sec": 1691566337
                },
                "status": {
                    "active": true,
                    "terminating": false
                }
            }
        },
        "metricset": {
            "name": "state_namespace",
            "period": 10000
        },
        "service": {
            "address": "127.0.0.1:55555",
            "type": "kubernetes"
        }
    }
]
```

Which will cause the conflict with the string field `kubernetes.namespace` @ChrsMark

Reply:
We don't need to change the `kubernetes.namespace_state` part. This can be as is.
I only suggest to change the value of `event.dataset`. I guess it's doable to just set this specifically within the code.

Reply:
We could try that, but how do I set that field? I had a look at the code and:

```go
e, err := util.CreateEvent(event, "kubernetes."+resourceName)

if err != nil {
	m.Logger().Error(err)
}

``` 

between these lines, there is still not `event.dataset` field.

The `event` is `{"_module":{"namespace":"kube-node-lease"},"created":{"sec":1693554126},"status":{"active":true,"terminating":false}}` (example) and `e` is `{{} {"namespace":"kube-node-lease"} {"created":{"sec":1693554126},"status":{"active":true,"terminating":false}}   kubernetes.namespace_state 0001-01-01 00:00:00 +0000 UTC <nil>   0s 0s false}` (example).

Reply:
I know that we have change our minds here so many times, just changing event.dataset = kubernetes.namespace_state, breaks also the user experience that metrics that come from kube state metrics have the pattern:`kubernetes.state_*`

Just retested those and wanted to bring this to discussion once more. @elastic/obs-cloudnative-monitoring team any thoughts?

Reply:
Maybe it's best to have this discussion over a meeting? Otherwise it seems it will take some time until we all align. @gizas @ChrsMark